1-fill-hpc-section (#7)
umr-amap/amap-dev-website: src/hpc.qmd

> ---

Reviewer: resources instead of ressources

title: "HPC"
execute:
  eval: false
---

# What is HPC ?

> High performance computing (HPC) is the practice of aggregating ("clustering") computing resources that work in parallel to gain performance greater than that of a single workstation and process massive multi-dimensional data sets.

Reviewer 2: ```suggestion
High performance computing (HPC) is the practice of aggregating ("clustering") computing resources that work in parallel to gain greater performance than a single workstation and process massive multi-dimensional data sets.
```

> A standard computing system solves problems primarily by using serial computing—it divides the workload into a sequence of tasks and then runs the tasks one after the other on the same processor. On the other hand, HPC uses parallel processing, which divides the workload into tasks that run simultaneously on multiple processors.

Reviewer: computing-it ça veut dire computing unit ?

Reviewer 2: Non, c'est un trait de séparation, un `:` serait plus clair. Je propose une reformulation : 

```suggestion
Computing system solves problems by dividing the workload into a subset of tasks. In a serial/sequential approach, tasks are executed one after the other on the same processor. In a parallel approach, the task are executed simultaneously/concurrently on mulitple processors.
```

HPC computer systems are characterized by their high-speed processing power, high-performance networks, and large-memory capacity.

While a laptop or desktop with a 3 GHz processor can perform around 3 billion calculations per second (~3.10^9), HPC solutions that can perform quadrillions of calculations per second (~10^{15}).

## Do you need HPC ?

> There are many cases where you might need to use use HPC resources.

Reviewer 2: ```suggestion
There are many cases where you might need to use HPC ressources.
```

Other options for HPC on the cloud (plateforms like AWS, Azure, Google Cloud) are also available, but are not covered here. These plateform are not free and are not eligible to any kind of support from POLIS.

### Meso-center `ISDM-Meso` {#sec-mesolr}
Hosted at [CINES](https://www.cines.fr/), this regional Meso-center [`Meso@LR`](https://meso-lr.umontpellier.fr/) provides access to HPC ressources for the Occitanie region. Although access to this ressource is not free, an existing agreement between CIRAD and ISDM makes it free for researchers and accessible upon request only (most likely within a certain limit of computing hours, but we have never been able to verify that :smile:).

#### Description of the cluster
Technical specificities include:

- 308 computational nodes of 28 cores each (bi processors Intel Xeon E5-2680 v4 2,4 Ghz (broadwell) 2×14 coeurs/noeud), 128 Go RAM
- 2 nodes with big memory (`bigmem`) with 112 coeurs, 3To RAM each

> - 2 noeuds with GPU and 52 CPU cores (bi-processors 26 cores) Poweredge R740 embedding RTX6000

Reviewer: nodes instead of noeuds

`Jean-Zay` cluster at IDRIS has recently been updated (09/2024) to the most powerfull `NVIDIA` GPUs available on the market (H100)

`Joliot-Curie` cluster at TGCC provides access `Nvidia V100 SXM2` GPUs and Intel CPUs. Note that you can also have access to `Quantique plateform`.

#### How to open an account ?
Anyone at AMAP can open an account to GENCI ressources. The process is a bit longer than for `Meso@LR` but is still quite fast (< 1 week) if you don't ask for too much ressources.

> warning
> Before submitting any project, we highly recommand you to contact the POLIS HPC-team at AMAP (currently T. Arsouze, P. Verley and D. Lamonica) to discuss your needs and the best way to use the ressources.

There are two ways to access GENCI ressources:

- "Accès dynamiques" (fast access) : if your needs are < 500 000 h CPU and 50 000 h GPU (eq. V100) for 1 year. This is a simplified access, only a one paragraph description of the scientific context is required and no scientific or technical review is done.

> - "Accès réguliers" (regular access) : if your needs are too important for dynamical access, then you must go for regular access. You can request for ressources only two time per year, and provide a detailed description of your project, both scientificaly and technicaly.

Reviewer: two times instead of two time

Reviewer: -lly instead of scientificaly and technicaly

You can find the "Quickconnect" bar at the top of the main window, and you need to provide the address of the "Host" : `muse-login.meso.umontpellier.fr`, your "Username" and "Password". You can also save your connection for later use : "File" -> "Copy current connection to Site Manager…".

##### Where to put my files ?
HPC clusters usually have multiple places where to put your files. Here are the ones for `Meso@LR`:

- `/home/my_login` : this is your home directory. You can put your scripts and small files here. This directory is backed up.
- `/lustre/scratch` : this is a temporary directory with very fast access where you can put your large files and data. This directory is not backed up and can be erased after 2 months. This space is recommended for running your computations that can require large amount of data.
- `/nfs/work` : this is a directory where you can put your input / output data or executable files. This directory is backed up and can be used to store your data for a long time.

### Create your computation environment

> HPC clusters usually provide a lot of scientific software and libraries that you can use. They are installed in a module system that you can load and unload. Some usefull commands are:

Reviewer: softwares instead of software, usefull instead of useful

```bash
$ cd ~       # back to your HOME directory
$ mkdir .R   # create a directory for R packages configuration
$ echo "CXX14 = g++ -fPIC\nCXX14FLAGS = -O3\nCXX14PICFLAGS = -fpic\nCXX14STD = -std=gnu++1y\n" > ~/.R/Makevars
# You can now install your packages
$ R --vanilla
> install.packages('brms')
Installation du package dans ‘/nfs/work/agap_id-bin/img/R/4.3.1-IFB/lib’
(car ‘lib’ n'est pas spécifié)
Avis dans install.packages("brms") :
  'lib = "/nfs/work/agap_id-bin/img/R/4.3.1-IFB/lib"' ne peut être ouvert en écriture
Voulez-vous plutôt utiliser une bibliothèque personnelle ? (oui/Non/annuler) oui
```

This will install your packages in the `$HOME/R/x86_64-pc-linux-gnu-library/4.3.1` directory.

- You can also use [conda](https://docs.conda.io/en/latest/) to create a dedicated environment and install additional packages. Although mostly used for installing python packages, conda can also be used to install R packages (search search for packages [here](https://anaconda.org/search?q=r-)).

### Submit jobs

> All HPC clusters use a job scheduler to manage the ressources and the jobs submitted by users. The most common job scheduler is `SLURM` (Simple Linux Utility for Resource Management). This is the scheduler use at `meso@lr`, but also àdastra` and `jean-zay`.

Reviewer: 'adastra' instead of àdastra

#### The SLURM scheduler
You can find the documentation for `SLURM` [here](https://slurm.schedmd.com/).

You can run your job in 2 modes:
- direct / interactive mode, using the command `srun` : for example `srun --pty bash` will open a bash session on a compute node.
- batch mode, using the command `sbatch` : you need to write a script that will be executed by the scheduler.

It is recommended to use the batch mode for long and heavy computations, and the interactive mode for testing your code.

#### How to write a SLURM script

> We refer to `SLURM` documentation to see all the options available, but we list here the most basic and usefull ones to write a batch script.

Reviewer: useful instead of usefull

|Slurm long format |	Slurm short format |	Description |
|-------------|:-------|:-------------------|
|--job-name	| -J |	Name of the job |
|--account | | Name of the account to which the job is charged |
|--nodes |	-N	| Number of nodes required |
|--ntasks	| -n	| Number of tasks of your job |
|--ntasks-per-node |  | Number of tasks per node (should be equal to the total number of tasks divided by the number of nodes) |

> |--ntasks-per-core |  | Number of tasks per core (ussually 1) |

Reviewer: usually instead of ussually

|--error	| -e |	Error file where the standard error will be written. By default, slurm-%j.err or '%j' is the job number |
|--mail-user |	|	Email address to send notifications |

#### How to submit and monitor a job
#### Different partitions available
Different types of ressources are available on the cluster, as listed in @sec-mesolr:

- __CPU__ node. Partition name : `agap_long`, `agap_normal`, `agap_short`

- __GPU__ node. Partition name : `agap_gpu`

- __Big memory__ node. Partition name : `agap_bigmem`

You can find the list of partitions available on the cluster with the command `sinfo` (including time limits and number of nodes available).

# Concrete examples
Here is a very simple script that you can adapt to your needs. You can fill additional `SLURM` options, and also the modify the call to your script.

```bash
#!/bin/bash                             # mandatory first line. This is a bash script
#SBATCH --job-name="<job_name>"         # name of the job
#SBATCH --partition=partition_name      # name of the partition you want to use
#SBATCH --nodes=1                       # number of nodes required

> #SBATCH --exclusive                     # Ask SLURM to reserve the whole nodes. This is not desired if you only need a few cores.

Reviewer: ask instead of Ask

> #SBATCH --time=1:00:00                  # Wall time requested (HH:MM:SS)

Reviewer: wall instead of Wall
